docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+
+```golang
+type DatasetConfig struct {
+
+    // One of the following can be set.
+    HFDatasetProvider *kubeflowv1.HFDatasetProvider `json:"huggingface,omitempty"`
+
+    S3DatasetProvider *kubeflowv1.S3DatasetProvider `json:"s3,omitempty"`
+
+    // (Optional). We can support the Iceberg dataset using PyIceberg.
+    IcebergDatasetProvider *kubeflowv1.IcebergDatasetProvider `json:"iceberg,omitempty"`
+}
+
+type HFDatasetProvider struct {
+  // Path to the HF dataset. For example: yelp-review-full
+  RepoId string `json:"repoId"`
+
+  // Whether the dataset needs to be splitted: train/val. E.g. split=train[:50000]
+  Split string `json:"split,omitempty"`
+
+  // Secret must contain HF_TOKEN secret.
+  // If the secret object contains more than one secret, all secrets are passed.
+  AccessTokenSecretRef corev1.SecretReference `json:"accessToken,omitempty"`
+}
+
+
+type S3DatasetProvider struct {
+  // S3 endpoint.
+  EndpointUrl string `json:"endpointUrl,omitempty"`
+
+  // Name of the S3 bucket.
+  BucketName string `json:bucketName`
+
+   // Path to the dataset. All files will be downloaded in that path.
+  Path string `json:path`
+
+  // Secret must contain AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY secret.
+  // If the secret object contains more than one secret, all secrets are passed.
+  // Otherwise, IRSA can be used to access S3.
+  AccessTokenSecretRef corev1.SecretReference `json:"accessToken,omitempty"`
+}
+```
+
+The following table explains how `TrainingRuntime` parameters will be overridden with the
+`DatasetConfig`.
+
+All parameters will be set for this container:
+
+```yaml
+.spec.replicatedJobs[name='Initializer'].template.spec.template.spec.containers[name=dataset-initializer]
+```
+
+#### The HuggingFace Provider
+
+For the HuggingFace provider environment variable `PROVIDER=hf`.
+
+<table>
+  <tr>
+   <td><strong><code>DatasetConfig</code> Parameter</strong>
+   </td>
+   <td><strong><code>TrainingRuntime</code> Parameter</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.huggingface.repoId</code>
+   </td>
+   <td><code>.env[REPO_ID]</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.huggingface.split</code>
+   </td>
+   <td><code>.env[SPLIT]</code>
+   </td>
+  </tr>
+</table>
+
+#### The S3 Provider
+
+For the S3 provider environment variable `PROVIDER=s3`.
+
+<table>
+  <tr>
+   <td><strong><code>DatasetConfig</code> Parameter</strong>
+   </td>
+   <td><strong><code>TrainingRuntime</code> Parameter</strong>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.s3.endpointUrl</code>
+   </td>
+   <td><code>.env[ENDPOINT_URL]</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.s3.bucketName</code>
+   </td>
+   <td><code>.env[BUCKET_NAME]</code>
+   </td>
+  </tr>
+  <tr>
+   <td><code>.s3.path</code>
+   </td>
+   <td><code>.env[PATH]</code>
+   </td>
+  </tr>
+</table>
+
+### The Model Config API
+
+The `ModelConfig` represents the APIs that data scientists can use to configure pre-trained model
```

Comment:
>This gets tricky very fast because not only are we now considering custom user preferences (e.g., "save my output to s3 under this path with these credentials" 

I think, you are right, since user also might want to configure some parameters to export model. 
We might want to design this after initial implementation.

In that case, maybe we should add this to the `ModelConfig`:
```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://models/trained-bert:v1
    env:
      - name: REGISTRY
        value: ghcr.io
```

However, with `startupPolicyOrder: InOrder` policy we can easily create the appropriate JobSet template for it.

```
apiVersion: kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
metadata:
  name: torch-tune-llama-7b
spec:
  numNodes: 1
  startupPolicy:
    startupPolicyOrder: InOrder
  replicatedJobs:
    - name: Initializer
      template:
        spec:
          template:
            spec:
              containers:
                - name: dataset-initializer
                  image: docker.io/kubeflow/dataset-initializer
                  env:
                    - name: DATASET_PROVIDER
                      value: hf
                    - name: REPO_ID
                      value: tatsu-lab/alpaca
                  volumeMounts:
                    - mountPath: /workspace/dataset
                      name: dataset-initializer
                - name: model-initializer
                  image: docker.io/kubeflow/model-initializer
                  env:
                    - name: MODEL_PROVIDER
                      value: hf
                    - name: REPO_ID
                      value: meta-llama/Llama-2-7b
                    - name: TRANSFORMER_TYPE
                      value: AutoModelForCausalLM
                  volumeMounts:
                    - mountPath: /workspace/model
                      name: model-initializer
              volumes:
                - name: dataset-initializer
                  persistentVolumeClaim:
                    claimName: dataset-initializer
                - name: model-initializer
                  persistentVolumeClaim:
                    claimName: model-initializer
    - name: Node
      template:
        spec:
          template:
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflow/llm-trainer
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: 29400
                    - name: TRANSFORMER_TYPE
                      value: AutoModelForCausalLM
                    - name: LORA_CONFIG
                      value: |
                        {"peft_type": "LORA", "r": 8, "lora_alpha": 16}
                  command:
                    - torchrun hf_llm_training.py
                  resources:
                    limits:
                      nvidia.com/gpu: 2
                  volumeMounts:
                    - mountPath: /workspace/dataset
                      name: dataset-initializer
                    - mountPath: /workspace/pre-trained-model
                      name: model-initializer
                    - mountPath: /workspace/adapters
                      name: model-exporter
              volumes:
                - name: dataset-initializer
                  persistentVolumeClaim:
                    claimName: dataset-initializer
                - name: model-initializer
                  persistentVolumeClaim:
                    claimName: model-initializer
                - name: model-exporter
                  persistentVolumeClaim:
                    claimName: model-exporter
    - name: Exporter
      template:
        spec:
          template:
            spec:
              containers:
                - name: model-exporter
                  image: docker.io/kubeflow/model-exporter
                  volumeMounts:
                    - mountPath: /workspace/adapters
                      name: model-exporter
              volumes:
                - name: model-exporter
                  persistentVolumeClaim:
                    claimName: model-exporter
```

Reply:
Dumb question - instead of this:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://models/trained-bert:v1
    env:
      - name: REGISTRY
        value: ghcr.io
```
Why would you not have:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://ghcr.io/models/trained-bert:v1
```

Or it could be `storageUri: oras://ghcr.io/models/trained-bert:v1` to indicate pushing to an OCI registry with ORAS. I'm wondering why the registry is a separate variable when it is part of the URI?

Reply:
Yeah, you are right!
Similar to this KEP: https://github.com/kubernetes/enhancements/pull/4642, we should set this as follows:
```
storageUri: oci://ghcr.io/models/trained-bert:v1
```

Reply:
Since this API is meant to hide the complexity from data scientists, the keys that can be overridden/customized could be exposed for simplicity. For instance, on TrainJob

Instead of
```
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://ghcr.io/models/trained-bert:v1
 ```
 
 could we have
 ```
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    transformerType: AutoModelForCausalLM   <--- Key that is overridden by user
  trainedModel:
    registry: quay.io    <---- This would change the storageUri to be oci://quay.io/models/trained-bert:v1
 ```
 
 Tracking down which environment variable needs to be modified can be hard. A subset of customizable keys with well-explained documentation could make the API more user-friendly.

Reply:
I like that overall suggestion - but I think I would still do:

```yaml
modelConfig:
 preTrainedModel:
   storageUri: hf://bert-base-uncased
   transformerType: AutoModelForCausalLM
 trainedModel:
   storageUri: oci://quay.io/models/trained-bert:v1
```

Reply:
In the idea of having environment variables `TRANSFORMER_TYPE` like attributes `transformerType`?
If yes, I liked the attributes idea but having a way to use environment variables gives us flexibility.
Maybe we should support both having a defined priority? Example: If we have `transformerType` and `TRANSFORMER_TYPE` the last defined in the YAML has the priority. This is like Linux, the last defined "env" will be used.

Another thing, may we support ConfigMap? 

I know this is a lot for now. Maybe for v2.0.0 we should use `transformerType` and start a new discussion in v2.x.x about having env vars, config maps, etc

Reply:
> I like that overall suggestion - but I think I would still do:
> 
> ```yaml
> modelConfig:
>  preTrainedModel:
>    storageUri: hf://bert-base-uncased
>    transformerType: AutoModelForCausalLM
>  trainedModel:
>    storageUri: oci://quay.io/models/trained-bert:v1
> ```

Yes. If the storageUri had been baked into the blueprint, overriding the registry field in the TrainJob would put in the new storageUri in the final JobSet. Alternatively, the user can provide the entire storageUri to be replaced.

Reply:
> Is the idea of having environment variables `TRANSFORMER_TYPE` like attributes `transformerType`? If yes, I liked the attributes idea but having a way to use environment variables gives us flexibility. Maybe we should support both having a defined priority? Example: If we have `transformerType` and `TRANSFORMER_TYPE` the last defined in the YAML has the priority. This is like Linux, the last defined "env" will be used.
> 
> Another thing, may we support ConfigMap?
> 
> I know this is a lot for now. Maybe for v2.0.0 we should use `transformerType` and start a new discussion in v2.x.x about having env vars, config maps, etc

Agree with the point on flexibility. Though, having to add an environment variable would mean data scientists understand the internal implementation, for ex: TRANSFORMER_TYPE is used as an environment variable and not a command line flag. The most useful config map keys could be similarly converted to attributes. Also, if there are a bunch of env variables (env or config map) that always get added in every trainJob, then it may make sense to create a blueprint with it, as per the overall intention of this proposal.

Reply:
The main problem with API attributes is that different model providers and dataset providers might have different configurations user can override. For example, S3 provider might have endpoint URL, but HF doesn't not have it.
Thus, this approach is making override more extensible to support various storage initializers:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    env:
      - name: TRANSFORMER_TYPE
        value: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://quay.io/models/trained-bert:v1
datasetConfig:
  storageUri: s3://datasets/yelp-review
  env:
    - name: ENDPOINT_URL
      value: s3.custom.com
```
Eventually, all of those env parameters should exist in the blueprint `model-initializer` and `storage-initializer` containers under `Initializer` Job.

@tenzen-y @shravan-achar @vsoch @diegolovison @johnugeorge @deepanker13 Any thoughts on how we can construct those API?

Reply:
If I understand the example from [this comment](https://github.com/kubeflow/training-operator/pull/2171#discussion_r1681926971), it seems like the `env` section is being used as a place to put random variables hat are relevant for the different Initializer jobs, and one example subset is related to the storageUri. I don't think I like (or see the need for) having redundancy there - more specifically any variable that can be represented in a Uri should be there, and it's easier to have one source of truth (vs. allowing an environment variable that goes into the Uri, and then the user provides both and it's not clear what the final result will be). In the case of variables for the Initializer that have no overlap (no other place to put them) then we are just using `env` as a storage place for key value pairs. In terms of what it looks like for the user, if we wanted something a bit more straight forward, why not just allow key value pairs, always string / string, that can be customized? Even storageURI could be one of them, e.g.,:

```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    transformerType: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://quay.io/models/trained-bert:v1
datasetConfig:
  storageUri: s3://datasets/yelp-review
  endpointUrl: s3.custom.com
```
Or if storageURI is special (or not a variable in the key/value set) make them explicitly separate:


```yaml
modelConfig:
  preTrainedModel:
    storageUri: hf://bert-base-uncased
    custom:
      transformerType: AutoModelForCausalLM
  trainedModel:
    storageUri: oci://quay.io/models/trained-bert:v1
datasetConfig:
  storageUri: s3://datasets/yelp-review
  custom:
    endpointUrl: s3.custom.com
```
And then the rule would be that anything under custom has a straight forward mapping to an environment variable, such as replacing uppercase or dashes with an underscore, and then making the entire thing uppercase. Then the underlying container would know how to use those. For the actual environment mapping, I think we want the final variable names to be namespaced. E.g.,:

```bash
DATASET_CONFIG_ENDPOINT_URI=s3.custom.com
MODEL_CONFIG_TRANSFORMER_TYPE=AutoModelForCasualLM
```
That way, I can do `env | grep PREFIX` to find an entire set for a particular type, and if necessary, they can exist in the same space together (where there could be overlapping names).

I can think more about the yaml design - I think I want to better understand cases (outside of storage URIs) these variables are used and needed.

Reply:
Yes, key-value pairs also should work. I know that KServe is doing this in the `storageSpec` API:
https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/predictor.go#L99-L101
Let me try to refine the API and we can discuss it again.

Reply:
This KEP has quite a lot in it. My gut is thinking that ModeConfig and DataSetConfig could be its own API that lives in Kubeflow/Common or somewhere else. 

I'd love to see the JobSet + TrainJob interface merged eventually and I worry that there will be a lot of back and forth on ModeConfig/DatasetConfig and that may block the progression of this feature.

I don't know where I found it but someone has already asked about this for pipelines and I'm sure notebooks would also want to use the same API. Maybe we consider splitting this from this for now?

Reply:
Yes, @diegolovison and @tarilabs discuss it here: https://github.com/kubeflow/training-operator/pull/2171#discussion_r1681096006.
We can always improve these API in the future version, that is why we call them `v2alpha1` initially.

I just think that the `datasetConfig` and `modelConfig` API will be very beneficial for us to work on, since majority of users are doing fine-tuning today.
Especially, this will be useful when user runs multi-note multi-worker training, when data needs to be distributed across multiple Pods.

About other Kubeflow components, I know that KServe is already using storage initializer to download model.

I don't think we will get any blockers from other Kubeflow WGs to implement it.

On the planning side, this APIs should be implemented after initial support of simple PyTorch Distributed Runtime.

Reply:
Sounds good to me. I tend to making KEPs/designs as simple as possible so we don't get bog down on details.